File: sdk/flutter/dazzle_flutter/ios/Classes/DazzleFlutterPlugin.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// iOS side of dazzle_flutter. Register the method channel, forward
// start/stop/waitForReady/isRunning calls to `DazzleServer.shared` —
// the exact same API the native Swift SDK consumers use.

import Flutter
import Foundation

@objc public class DazzleFlutterPlugin: NSObject, FlutterPlugin {

    // Retained for the lifetime of the plugin so the bridge's channel
    // delegates don't get deallocated mid-stream.
    private static var fmBridge: AnyObject?
    private static var anthropicBridge: AnyObject?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "dev.dazzle.flutter",
            binaryMessenger: registrar.messenger()
        )
        let instance = DazzleFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Apple Foundation Models bridge — iOS/macOS 26+ only. On
        // older OSes we register a stub that replies `false` to the
        // `isAvailable` probe so Dart-side callers fall back cleanly.
        if #available(iOS 26.0, macOS 26.0, *) {
            fmBridge = FoundationModelsBridge(messenger: registrar.messenger())
        } else {
            let stub = FlutterMethodChannel(
                name: "dev.dazzle.flutter/foundation",
                binaryMessenger: registrar.messenger())
            stub.setMethodCallHandler { call, result in
                if call.method == "isAvailable" { result(false) }
                else { result(FlutterMethodNotImplemented) }
            }
        }

        // Anthropic bridge — works on all iOS versions; dormant
        // until the Dart-side `AnthropicClient` constructor invokes
        // `create`. Uses `URLSession` directly, no extra deps.
        anthropicBridge = AnthropicBridge(messenger: registrar.messenger())
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":      handleStart(call, result)
        case "stop":       handleStop(result)
        case "waitForReady": handleWaitForReady(call, result)
        case "isRunning":  result(DazzleServer.shared.isRunning)
        default:           result(FlutterMethodNotImplemented)
        }
    }

    // MARK: – Method handlers

    private func handleStart(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        do {
            if DazzleServer.shared.isRunning {
                result(nil); return
            }
            let cfg = parseConfig(call.arguments as? [String: Any])
            try DazzleServer.shared.start(config: cfg)
            result(nil)
        } catch {
            result(FlutterError(
                code: "DAZZLE_START_FAILED",
                message: "\(error)",
                details: nil
            ))
        }
    }

    private func handleStop(_ result: @escaping FlutterResult) {
        if DazzleServer.shared.isRunning {
            DazzleServer.shared.stop()
        }
        result(nil)
    }

    private func handleWaitForReady(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let timeoutMs = args?["timeoutMs"] as? Int ?? 5000
        let ok = DazzleServer.shared.waitForReady(timeout: TimeInterval(timeoutMs) / 1000.0)
        result(ok)
    }

    // MARK: – Config parsing

    private func parseConfig(_ m: [String: Any]?) -> DazzleConfig {
        guard let m = m else { return DazzleConfig() }

        let maxMemory = (m["maxMemory"] as? String) ?? "64mb"

        let modules: Set<DazzleModule> = {
            let arr = m["modules"] as? [String] ?? []
            var set = Set<DazzleModule>()
            if arr.contains("vectorSearch") { set.insert(.vectorSearch) }
            return set
        }()

        let wipe: WipeTarget = {
            let arr = m["wipeOnStart"] as? [String] ?? []
            var opts: WipeTarget = []
            if arr.contains("aof") { opts.insert(.aof) }
            if arr.contains("rdb") { opts.insert(.rdb) }
            return opts
        }()

        let persistence: DazzlePersistence = {
            guard let p = m["persistence"] as? [String: Any] else { return .aof() }
            switch p["kind"] as? String {
            case "none": return .none
            case "rdb":  return .rdb()
            default:
                let fsync: AppendFsync
                switch p["fsync"] as? String {
                case "always": fsync = .always
                case "no":     fsync = .no
                default:       fsync = .everysec
                }
                return .aof(fsync: fsync)
            }
        }()

        return DazzleConfig(
            maxMemory:   maxMemory,
            persistence: persistence,
            wipeOnStart: wipe,
            modules:     modules
        )
    }
}

File: sdk/flutter/dazzle_flutter/android/src/main/kotlin/dev/dazzle/flutter/DazzleFlutterPlugin.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// Android side of the dazzle_flutter plugin.
//
// Responsibilities:
//   1. Register the `dev.dazzle.flutter` method channel so the Dart
//      `DazzleServer` can drive the embedded Valkey server's lifecycle
//      (start, stop, waitForReady, isRunning).
//   2. Forward those calls to the real Dazzle SDK's `DazzleServer`
//      object — the same one native Kotlin apps use.
//
// All DATA operations (HSET/HGETALL/ZADD/FT.SEARCH/…) skip this plugin
// entirely and go directly from Dart dart:ffi → libdazzle.so → the
// in-process transport. The method channel exists only for the
// lifecycle ceremony.

package dev.dazzle.flutter

import android.content.Context
import dev.dazzle.sdk.AppendFsync
import dev.dazzle.sdk.DazzleConfig
import dev.dazzle.sdk.DazzleModule
import dev.dazzle.sdk.DazzlePersistence
import dev.dazzle.sdk.DazzleServer
import dev.dazzle.sdk.WipeTarget
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class DazzleFlutterPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var liteRtBridge: LiteRtBridge? = null
    private var anthropicBridge: AnthropicBridge? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "dev.dazzle.flutter")
        channel.setMethodCallHandler(this)
        // LiteRT-LM bridge — sits on its own method + event channels.
        // Dead code unless the consumer actually adds the
        // `com.google.ai.edge.litertlm:litertlm-android` runtime to
        // their app module — the bridge class resolves at first
        // `create` call, and LiteRtLmClient's init fails early with
        // a clear message if the runtime is missing.
        liteRtBridge = LiteRtBridge(context, binding.binaryMessenger)
        // Anthropic bridge — also dormant until the Dart-side
        // `AnthropicClient` constructor runs `create` on it. No
        // runtime deps; uses the Kotlin SDK's `AnthropicClient`
        // (HttpURLConnection + org.json, zero external libs).
        anthropicBridge = AnthropicBridge(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        liteRtBridge?.dispose()
        liteRtBridge = null
        anthropicBridge?.dispose()
        anthropicBridge = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> handleStart(call, result)
            "stop" -> {
                if (DazzleServer.isRunning()) DazzleServer.stop()
                result.success(null)
            }
            "waitForReady" -> {
                // DazzleServer.start() is synchronous on JVM — once it
                // returns, the server is ready. For parity with the
                // Dart-side API we still spin-poll isRunning() until
                // the given deadline.
                val timeoutMs = (call.argument<Int>("timeoutMs") ?: 5000).toLong()
                val deadline = System.currentTimeMillis() + timeoutMs
                var ok = DazzleServer.isRunning()
                while (!ok && System.currentTimeMillis() < deadline) {
                    Thread.sleep(25)
                    ok = DazzleServer.isRunning()
                }
                result.success(ok)
            }
            "isRunning" -> result.success(DazzleServer.isRunning())
            else -> result.notImplemented()
        }
    }

    /**
     * Parse the DazzleConfig map the Dart side sent and boot the server
     * exactly as a native Kotlin caller would. Idempotent — calling
     * `start` when the server is already running is a no-op.
     */
    private fun handleStart(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (DazzleServer.isRunning()) {
                result.success(null)
                return
            }
            @Suppress("UNCHECKED_CAST")
            val m = (call.arguments as Map<String, Any?>?)
            val cfg = parseConfig(m)
            DazzleServer.start(context, cfg)
            result.success(null)
        } catch (t: Throwable) {
            result.error("DAZZLE_START_FAILED",
                "${t::class.simpleName}: ${t.message}", null)
        }
    }

    private fun parseConfig(m: Map<String, Any?>?): DazzleConfig {
        if (m == null) return DazzleConfig()

        val port      = (m["port"] as? Int) ?: 0
        val maxMemory = (m["maxMemory"] as? String) ?: "64mb"
        val modules   = (m["modules"] as? List<*>)
            ?.mapNotNull { name ->
                when (name) {
                    "vectorSearch" -> DazzleModule.VectorSearch
                    else -> null
                }
            }?.toSet() ?: emptySet()
        val wipeOnStart = (m["wipeOnStart"] as? List<*>)
            ?.mapNotNull { name ->
                when (name) {
                    "aof" -> WipeTarget.AOF
                    "rdb" -> WipeTarget.RDB
                    else  -> null
                }
            }?.toSet() ?: emptySet()

        val persistence = parsePersistence(
            m["persistence"] as? Map<*, *>
        )

        return DazzleConfig(
            port        = port,
            maxMemory   = maxMemory,
            persistence = persistence,
            wipeOnStart = wipeOnStart,
            modules     = modules,
        )
    }

    private fun parsePersistence(m: Map<*, *>?): DazzlePersistence {
        if (m == null) return DazzlePersistence.Aof()
        return when (m["kind"]) {
            "none" -> DazzlePersistence.None
            "rdb"  -> DazzlePersistence.Rdb()  // Saves list passed through separately by users who need it.
            else -> {
                val fsync = when (m["fsync"] as? String) {
                    "always"   -> AppendFsync.ALWAYS
                    "no"       -> AppendFsync.NO
                    else       -> AppendFsync.EVERYSEC
                }
                DazzlePersistence.Aof(fsync)
            }
        }
    }
}
